### ISSUE TITLE
Error: Request Failed with status 400 - Bad Request

### ISSUE THREAD

**Issue by dkhan007** [BigThunderSR/node-red-contrib-onstar2#271]
As always, I really am grateful for your modules.

It is more of a question, as I am getting error 400, I've checked my password, pin, VIN and generated uuid...all seem ok to me.  I've uninstall the node, restarted home assistant, and reinstalled and still get the same issue. 

Additionally what's odd, is that the error comes right away when the node is run, as opposed to a few seconds later after it's communicating with Onstar.  Perhaps Onstar has blocked the API again?  Do you have any ideas that I can try to trouble shoot it on my end?

Thx for any help you can provide!

Node Red 18.1.1
Home Asssitant:
Core 2024.10.4
Supervisor 2024.10.3
Operating System 13.2

**Comment by dkhan007** [comment]
Ah got you.  I thought I was going crazy.  well... hopefully there is a way forward, if not, thx for the original node flows, it was great while it lasted.

**Comment by LightningManGTS** [comment]
For this part of the conversation,

@BigThunderSR 

From the work done here: https://github.com/samrum/OnStarJS/issues/233 how much of what's in OnStarJS is needed that isn't being taken care of with what is in the rewrite joelvandal and methos did? Is the way they implement it with nodejs an issue or could this project be reconjigured to use this code instead? https://github.com/joelvandal/node-oauth2-gm

**Comment by LightningManGTS** [comment]
Thanks for the update @BigThunderSR! 

After adding my TOTP secret code to my cars config I'm getting a missing require configuration parameter error.

**Comment by LightningManGTS** [comment]
> > Thanks for the update @BigThunderSR!
> > After adding my TOTP secret code to my cars config I'm getting a missing require configuration parameter error.
> 
> I just noticed that it works fine in my standalone NodeRED instance, but it is failing in my HA addon NodeRED instance. Are you running NodeRED in HA or is it independent of HA?

Yeah I'm running it in HA. I have it configured through HA to install the package so all I ever have to do is reestart it from there to get the update. For reference I did finally get around to wrapping my head around MQTT just now and got that version of it working for onstar. No issues there which is awesome!

**Comment by jtmrebels** [comment]
Version 2.3.0 in HA Node-Red was working great for me all day after I added my TOPT secret. I updated tonight to 2.3.1 and now I am getting the "Error: Failed to get authorization code." I added a new TOPT secret but still the same error.

**Comment by jtmrebels** [comment]
> > Version 2.3.0 in HA Node-Red was working great for me all day after I added my TOPT secret. I updated tonight to 2.3.1 and now I am getting the "Error: Failed to get authorization code." I added a new TOPT secret but still the same error.
> > If it's helpful, I am getting the error below in MQTT:
> > error: Main function error: {"error":{"message":"Failed to extract csrf token or transId during MFA"},
> 
> Thanks for letting me know. I'll try to see what the issue is this weekend since it's happening on my system as well.
> 
> I'm not having any issues with either the standalone or HA add-on versions of onstar2mqtt though.

Thank you!! I appreciate all the support.

**Comment by eric-ewilson** [comment]
I updated to the new version, entered my TOTP secret, and now I'm getting "Error: EROFS: read-only file system, open 'microsoft_tokens.json'". I haven't seen this error before and I'm not sure where to begin troubleshooting.

**Comment by eric-ewilson** [comment]
It runs on a Mac Mini. On Jan 21, 2025, at 7:13 PM, BigThunderSR ***@***.***> wrote:﻿

I updated to the new version, entered my TOTP secret, and now I'm getting "Error: EROFS: read-only file system, open 'microsoft_tokens.json'". I haven't seen this error before and I'm not sure where to begin troubleshooting.

Mine is working in both a standalone NodeRED install on Ubuntu and also in the official Home Assistant NodeRED add-on. How are you running NodeRED?

**Comment by eric-ewilson** [comment]
I checked the .node-red directory in my user folder and all the
directories/subdirectories are writable by the user running NodeRED (me). I
also can't find microsoft_tokens.json on the computer. Any idea where that
file is supposed to be located?

**Comment by eric-ewilson** [comment]
Thanks. Your reply prompted me to manually start NodeRED so I could read
the debug output. In doing so, I learned that Onstar2 was able to function
properly. I had been using PM2 to automatically start NodeRED and PM2 seems
to have been the problem, even though it also runs under my user name. I
don't fully understand the problem since NodeRED was able to write files in
other flows in the original configuration. I solved the problem by
switching from PM2 to a LaunchAgent to start NodeRED.
